### ISSUE TITLE
tuples are comparable even when they would throw an error at runtime

### ISSUE THREAD

**Issue by indigoviolet** [microsoft/pyright#8901]
**Describe the bug**

If you have two variables with type `str | None`, pyright (and mypy) will correctly say they cannot be compared. But `tuple[str, str | None]` will not raise an error.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

Code sample in [pyright playground](https://pyright-play.net/?code=JYAgvCAMBQ0CYFMBmIDOB7ATgFwPoGsEBPACgA8AuNbTAShAFoA%2BEbAVwAcAbBAbVRoAaaphAAfEADl0AOwQBdCtBAqQAcy7oARgEMuIYMtWgA1BACMRlcBSgApCABM4CJCWqPITAnaYZIcmFpOVorEAQuVAR3TxVvX39AkDJQ2B1wNCw8QlIAIh1c0K0MjBwCYhJ8wtgALwz0gB4QLSA)

```python
i = 0

def sort_key(x: str) -> tuple[str, str | None]:
    global i
    i += 1
    if i % 2 == 0:
        return (x, None)
    else:
        return (x, x)


a = sort_key("a")
b = sort_key("a")

z = a < b
```

**VS Code extension or command-line**

playground, vscode/pylance, as well as pyright CLI (1.1.379)

**Comment by erictraut** [comment]
From the perspective of type checking, pyright (and mypy) are working correctly here. The relevant definition for `tuple.__lt__` can be found [here](https://github.com/python/typeshed/blob/236ec8e9cf1abcf7ca418445a6424f05ee07919f/stdlib/builtins.pyi#L955).

I don't see a good way to express this runtime behavior of `tuple.__lt__`, at least currently given the tools provided by the type system.

It would be theoretically possible for a type checker to provide custom logic that overrides the type definition in typeshed, but that's generally not advisable. I've done it a few times in the past with pyright, and I've come to regret it every time because it leads to maintenance headaches and hard-to-explain and inconsistent behaviors.

I'm going to close the issue because I don't think there's anything actionable here.
